lib/core/src/Cardano/Wallet.hs

```diff
@@ -2014,9 +2015,52 @@ joinStakePool ctx currentEpoch knownPools pid poolStatus wid argGenChange pwd =
         let action = if isKeyReg then Join pid else RegisterKeyAndJoin pid
         liftIO $ traceWith tr $ MsgIsStakeKeyRegistered isKeyReg
 
-        selection <- withExceptT ErrJoinStakePoolSelectCoin $
+        cs <- withExceptT ErrJoinStakePoolSelectCoin $
             selectCoinsForDelegation @ctx @s @t @k ctx wid action
 
+        cp <- mapExceptT atomically
+            $ withExceptT ErrJoinStakePoolNoSuchWallet
+            $ withNoSuchWallet wid
+            $ readCheckpoint (PrimaryKey wid)
+        let s = getState cp
+            dprefix = Seq.derivationPrefix s
+            sPath = stakePath dprefix
```

Comment:
Instead of fetching _just_ the metadata with `readWalletMeta`, you could fetch the wallet state + metadata, but more importantly, do the coin selection and post-processing within the same `atomically`. Otherwise, there might be another thread interleaving changing the state of the wallet after you ran the coin selection. This is pretty minor here I recall because you're not writing anything back, but that's a potential source of subtle bug. 

Said differently, we should avoid:

```
atomically readCheckpoint
// stuff
atomically writeCheckpoint / readCheckpoint again
// stuff
```

but instead do

```
atomically $ do
   readCheckpoint
   // stuff
   writeCheckpoint 
   // stuff
```

Reply:
Does this fix it?
```hs
joinStakePoolUnsigned' ctx currentEpoch knownPools pid poolStatus wid =
    db & \DBLayer{..} -> do
        (wal, walMeta, _) <- withExceptT
            ErrJoinStakePoolNoSuchWallet (readWallet @ctx @s @k ctx wid)
        isKeyReg <- mapExceptT atomically
            $ withExceptT ErrJoinStakePoolNoSuchWallet
            $ isStakeKeyRegistered (PrimaryKey wid)

        let mRetirementEpoch = view #retirementEpoch <$>
                W.getPoolRetirementCertificate poolStatus
        let retirementInfo =
                PoolRetirementEpochInfo currentEpoch <$> mRetirementEpoch

        withExceptT ErrJoinStakePoolCannotJoin $ except $
            guardJoin knownPools (walMeta ^. #delegation) pid retirementInfo

        let action = if isKeyReg then Join pid else RegisterKeyAndJoin pid
        liftIO $ traceWith tr $ MsgIsStakeKeyRegistered isKeyReg

        cs <- withExceptT ErrJoinStakePoolSelectCoin $
            selectCoinsForDelegation @ctx @s @t @k ctx wid action

        let s = getState wal
            dprefix = Seq.derivationPrefix s
            sPath = stakeDerivationPath dprefix

        pure (cs, action, sPath)
```